### DISCUSSION TITLE
Calling POST REST APIs (nodejs axios) will result to "The CSRF session token is missing"

### DISCUSSION THREAD

**Discussion by zedlauzy** [apache/superset#18373] [Q&A / Help] [answered]
Hello there, I am unable to do any POST api calls (import_dashboard) as it will lead to the 400 error `The CSRF session token is missing` using nodejs & axios. I do not wish to exempt these APIs in `WTF_CSRF_EXEMPT_LIST` if it is possible.

Edit: I am currently using Superset 1.2 with `VERSIONED_EXPORT` included in `FEATURE_FLAGS`

**Steps to Reproduce:**
1)  In my nodejs app, I have this axios setup where it imports dashboard via `importContent`:

```
const config = {
  headers: {
    'X-CSRFToken': await this.getCsrfToken(),
    'Authorization': `Bearer ${await this.getAccessToken()}`
  }
}

private async getAccessToken(): Promise<string> {
  const loginRequest = this.createLoginRequest()
  const response = await axios.post(`${this.baseURL}api/v1/security/login`, loginRequest)
  return response.data['access_token']
}

private async getCsrfToken(): Promise<string> {
  const token = `Bearer ${await this.getAccessToken()}`

  const options: AxiosRequestConfig = {
    headers: {
      Authorization: token
    }
  }
  const url = `${this.baseURL}api/v1/security/csrf_token`
  const response = await axios.get(url, options)
  return response.data['result']
}

async importContent(contentType: string, filePath: string, overwrite: boolean): Promise<any> {
  const formData = new FormData()
  formData.append('formData', fs.createReadStream(filePath))
  formData.append('overwrite', String(overwrite))
  const options = {
      headers: {
        Authorization: `Bearer ${await this.getAccessToken()}`,
        'X-CSRFToken': await this.getCsrfToken(),
        ...formData.getHeaders()
      }
    }
  const url = `${this.baseURL}api/v1/${contentType}/import/`

  const response = await axios.post(url, formData, options)
  if (response.data) {
    return response
  }
}
```

2) I am sure the code is working fine as I have tried it with `WTF_CSRF_EXEMPT_LIST` having the import_dashboard api

3) It only goes wrong when I remove it from `WTF_CSRF_EXEMPT_LIST`. I also have verified via Postman with the similar request I have below is working:
```
Headers: {
  Authorization: 'Bearer {token}',
  X-CSRFToken: {csrfToken}
}
form-data: {
  formData: {uploadedFile}
}
```

**Actual Result:**
Getting `The CSRF session token is missing` error

**Expected Result:**
Importing dashboard via REST API call should work as expected

**Comment by zedlauzy** [comment, marked as answer]
Hi there, can I get some advice? I have updated to the latest `1.4.1` superset but the same issue is still recurring. Can I check if this error is possible to overcome without adding the APIS in `WTF_CSRF_EXEMPT_LIST`?

**Reply by ad-m** [reply]
Yes, you can support work in #18123, then extend that work to JavaScript (if JavaScript is must have). It will allows overcome without adding the APIS in `WTF_CSRF_EXEMPT_LIST`

**Reply by zedlauzy** [reply]
Thanks for clarifying. I assume exempting is the only solution I have now
